Consolidate contribution guidelines. (#10689)
hyperledger/besu: CONTRIBUTING.md

Please follow these steps to have your contribution considered by the approvers:

1. Ensure all commits have a sign-off for the DCO, as described in [Developer Certificate of Origin (DCO)](#developer-certificate-of-origin-dco).
2. Follow all instructions in the [pull request template](https://github.com/besu-eth/besu/blob/main/.github/pull_request_template.md).
3. Include appropriate test coverage. Testing is 100% automated; all submissions must be testable in an automated fashion.
4. Follow the coding conventions in the [Developing and Conventions](https://github.com/besu-eth/besu/wiki/Developing-and-Conventions) wiki page.
5. After you submit your pull request, verify that all [status checks](https://docs.github.com/articles/about-status-checks) are passing.

### What makes a good pull request?

#### One pull request, one change

- This limits the surface area of the change and makes it easier to identify root causes when issues arise.
- Make sure your PR doesn't include commits that are not part of it. This can happen if [your fork is not up to date](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/syncing-a-fork).

#### Link to an issue

> - When submitting your PR, include the issue's link in the description and the number in the title (for example, `[#99] My awesome PR`). This provides context and auto-updates the issue to link to your PR.

Reviewer: this is duplicated content, can we remove

Author: Fixed!

2. **Clone your fork** to your computer.
3. **Create a topic branch** and name it appropriately. Starting the branch name with the issue number is a good practice and a reminder to fix only one issue per PR.
4. **Make your changes** adhering to the coding conventions described in the [Besu wiki](https://github.com/besu-eth/besu/wiki). In general a commit serves a single purpose, and diffs should be easily comprehensible. For this reason, do not mix formatting fixes or code moves with actual code changes.
5. **Commit your changes**. See the [How to Write a Git Commit Message](https://cbea.ms/git-commit/) article by Chris Beams. Make sure to add a DCO sign-off to each commit (for example, `git commit -s -m "..."`); see [Developer Certificate of Origin (DCO)](#developer-certificate-of-origin-dco) below.
6. **Test your changes** locally before pushing to ensure you are not breaking another part of the software. Running `./gradlew clean check test` locally helps you be confident that your changes will pass CI once pushed as a PR.
7. **Push your changes** to your remote fork (usually labeled `origin`).

> 8. **Create a pull request** on the Besu repository. If it's not ready for review, make it a `Draft` PR. If the PR addresses an existing issue, include the issue number in the PR title in square brackets (for example, `[#2374]`).

Reviewer: I actually think including the issue number like that in PR title is confusing. GH UI does an excellent job of linking PRs and issues together, I think refs or closes in the body of the PR is sufficient - interested to hear other folks thoughts

Reviewer 2: Agree, I don't think there is any need to do that in the title. If people include the #fix with the PR that it links to that should be enough IMHO.

Author: fixed.

- **A description of the current behavior** and an explanation of which behavior you expected to see instead, and why.
- **An explanation of why this enhancement would be useful** to most users.
- **Whether this enhancement exists in other clients.**
- **The version of Besu you're using** (`besu --version`) and the name and version of your OS.

## Pull requests

The pull request process has several goals:

- Maintain product quality.
- Fix problems that are important to users.
- Engage the community in working toward the best possible product.
- Enable a sustainable system for maintainers to review contributions.

Please follow these steps to have your contribution considered by the approvers:

1. Ensure all commits have a sign-off for the DCO, as described in [Developer Certificate of Origin (DCO)](#developer-certificate-of-origin-dco).
2. Follow all instructions in the [pull request template](https://github.com/besu-eth/besu/blob/main/.github/pull_request_template.md).
3. Include appropriate test coverage. Testing is 100% automated; all submissions must be testable in an automated fashion.

> 4. Follow the coding conventions and [style guides](https://github.com/besu-eth/besu/wiki).

Reviewer: this link is missing something

5. After you submit your pull request, verify that all [status checks](https://docs.github.com/articles/about-status-checks) are passing.

### What makes a good pull request?

#### One pull request, one change

- This limits the surface area of the change and makes it easier to identify root causes when issues arise.
- Make sure your PR doesn't include commits that are not part of it. This can happen if [your fork is not up to date](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/syncing-a-fork).

#### Link to an issue

- When submitting your PR, include the issue's link in the description and the number in the title (for example, `[#99] My awesome PR`). This provides context and auto-updates the issue to link to your PR.

#### Minimize lines of code (LOC) per PR

- PRs get near-exponentially longer to review as the number of lines of code increases. Ideally, keep your changes under 300 LOC. If that's not possible, try to break your PR into smaller ones for reviewers to review sequentially.

#### Write meaningful commit messages

> - Your commit title should include the issue number, and the description should link to the issue and comprehensively describe the changes.

Reviewer: ```suggestion
- Your commit messages should be meaningful, and the PR description should link to the related issue, and comprehensively describe the changes.
```

#### Be responsive

- Don't let a PR sit idle with unaddressed comments until it needs a full rebase. If you are pausing work on an issue, indicate it in the PR comments or change the PR to draft status.

#### What if the status checks are failing?

- If a status check is failing and you believe the failure is unrelated to your change, leave a comment on the pull request explaining why. A maintainer will re-run the status check for you. If we conclude that the failure was a false positive, we will open an issue to track the problem with our status check suite.

## Code reviews

> All changes must be code reviewed. For non-approvers this is obvious, since you can't commit anyway. But even for approvers, we want all changes to get at least one review, preferably (and, for non-trivial changes, obligatorily) from someone who knows the areas the change touches. For non-trivial changes we may want two reviewers. The primary reviewer makes this decision and nominates a second reviewer if needed. Except for trivial changes, PRs should not be committed until relevant parties (for example, owners of the affected subsystem) have had a reasonable chance to look at the PR in their local business hours.

Reviewer: ```suggestion
All changes must be code reviewed, preferably (and, for non-trivial changes, obligatorily) from someone who knows the areas the change touches. For non-trivial changes we may want two reviewers. The primary reviewer makes this decision and nominates a second reviewer if needed. Except for trivial changes, PRs should not be merged until relevant parties (for example, owners of the affected subsystem) have had a reasonable chance to look at the PR in their local business hours.
```

> Most PRs will find reviewers organically. If an approver intends to be the primary reviewer of a PR, they should set themselves as the assignee on GitHub and say so in a reply to the PR. Only the primary approver of a change should do the merge, except in rare cases (for example, they are unavailable in a reasonable timeframe).

Reviewer: ```suggestion
Most PRs will find reviewers organically. If an approver intends to be the primary reviewer of a PR, they should set themselves as the assignee on GitHub. Only the primary approver of a change should do the merge, except in rare cases (for example, they are unavailable in a reasonable timeframe).
```

> If a PR has gone two working days without an approver emerging, please ask on [Discord](https://discord.com/invite/hyperledger).

Reviewer: ```suggestion
If a PR has gone five working days without an reviewer emerging, you can ask on [Discord](https://discord.com/invite/hyperledger), however please don't ping or message individual maintainers.
```
